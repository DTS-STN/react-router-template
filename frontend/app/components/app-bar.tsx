import type { JSX } from 'react';

import { faChevronDown, faRightFromBracket, faUser } from '@fortawesome/free-solid-svg-icons';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { useTranslation } from 'react-i18next';

import { DropdownMenu, DropdownMenuContent, DropdownMenuLabel, DropdownMenuTrigger } from '~/components/dropdown-menu';
import { AppLink } from '~/components/links';
import { MenuItem } from '~/components/menu';
import { useLanguage } from '~/hooks/use-language';
import { cn } from '~/utils/tailwind-utils';

type AppBarProps = {
  name?: string;
};

export function AppBar({ name }: AppBarProps): JSX.Element {
  const { t } = useTranslation(['gcweb']);
  const { MSCA_BASE_URL } = globalThis.__appEnvironment;

  return (
    <div className="bg-slate-700">
      <div className="align-center container mx-auto flex flex-wrap justify-between">
        <div className="align-center flex">
          <span id="menu-label" className="my-auto py-2 text-white sm:text-2xl">
            <AppLink to={t('gcweb:app.menu-dashboard.href', { baseUri: MSCA_BASE_URL })} className="hover:underline">
              {t('gcweb:app.title')}
            </AppLink>
          </span>
        </div>
        <div className="flex items-center space-x-4 text-right">{name && <UserButton name={name} />}</div>
      </div>
    </div>
  );
}

type UserButtonProps = {
  className?: string;
  name?: string;
};

function UserButton({ className, name }: UserButtonProps): JSX.Element {
  const { t } = useTranslation(['gcweb']);
  const { currentLanguage } = useLanguage();
  const { MSCA_BASE_URL } = globalThis.__appEnvironment;

  return (
    <DropdownMenu>
      <DropdownMenuTrigger
        className={cn(
          'flex h-full flex-nowrap space-x-2 bg-slate-200 px-2 text-sm text-slate-700 hover:bg-neutral-300 hover:underline focus:bg-neutral-300 focus:ring-2 focus:ring-black focus:ring-offset-2 focus:outline-hidden aria-expanded:bg-neutral-300 aria-expanded:text-slate-700 sm:space-x-4 sm:px-4',
          className,
        )}
      >
        <div className="text-md my-auto flex flex-nowrap items-center space-x-2 py-2">
          <div className="flex h-8 w-8 items-center justify-center rounded-full bg-slate-700">
            <FontAwesomeIcon icon={faUser} className="size-5 text-slate-200" />
          </div>
          <span id="menu-label" className="text-md hidden py-2 font-bold sm:block">
            {name}
          </span>
        </div>
        <FontAwesomeIcon icon={faChevronDown} className="my-auto size-4" />
      </DropdownMenuTrigger>
      <DropdownMenuContent align="end" className="w-64">
        <UserName name={name} />
        <MenuItem
          to={t('gcweb:app.menu-dashboard.href', { baseUri: MSCA_BASE_URL })}
          className="text-md flex justify-between text-black hover:bg-zinc-100 hover:text-black focus:bg-zinc-100 active:bg-zinc-100"
        >
          {t('gcweb:app.menu-dashboard')}
        </MenuItem>
        <MenuItem
          to={t('gcweb:app.profile.href', { baseUri: MSCA_BASE_URL })}
          className="text-md flex justify-between text-black hover:bg-zinc-100 hover:text-black focus:bg-zinc-100 active:bg-zinc-100"
        >
          {t('gcweb:app.profile')}
        </MenuItem>
        <MenuItem
          to={t('gcweb:app.security-settings.href', { baseUri: MSCA_BASE_URL })}
          className="text-md flex justify-between text-black hover:bg-zinc-100 hover:text-black focus:bg-zinc-100 active:bg-zinc-100"
        >
          {t('gcweb:app.security-settings')}
        </MenuItem>
        <MenuItem
          to={t('gcweb:app.contact-us.href', { baseUri: MSCA_BASE_URL })}
          className="text-md flex justify-between text-black hover:bg-zinc-100 hover:text-black focus:bg-zinc-100 active:bg-zinc-100"
        >
          {t('gcweb:app.contact-us')}
        </MenuItem>
        <div role="separator" aria-orientation="horizontal" className="-mx-1 my-1 h-px bg-slate-100"></div>
        <MenuItem
          to={`/auth/logout?lang=${currentLanguage}`}
          className="text-md flex justify-between text-black hover:bg-zinc-100 hover:text-black focus:bg-zinc-100 active:bg-zinc-100"
        >
          {t('gcweb:app.logout')}
          <FontAwesomeIcon icon={faRightFromBracket} className="my-auto size-8" />
        </MenuItem>
      </DropdownMenuContent>
    </DropdownMenu>
  );
}

type UserNameProps = {
  name?: string;
};

function UserName({ name }: UserNameProps): JSX.Element {
  return (
    <>
      {name !== undefined && (
        <DropdownMenuLabel className="text-md flex items-center border-b-2 border-slate-600 px-3 py-2 text-gray-300 sm:hidden">
          <FontAwesomeIcon icon={faUser} className="mr-2 size-4" />
          {name}
        </DropdownMenuLabel>
      )}
    </>
  );
}
