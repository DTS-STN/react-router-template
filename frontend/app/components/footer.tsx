import { useTranslation } from 'react-i18next';

import { AppLink } from './links';

export interface FooterProps {
  bilingual: boolean;
}

export function Footer({ bilingual }: FooterProps) {
  const { i18n, t } = useTranslation(['gcweb']);
  const en = i18n.getFixedT('en');
  const fr = i18n.getFixedT('fr');
  const { MSCA_BASE_URL, ECAS_BASE_URL } = globalThis.__appEnvironment;

  return (
    <footer id="wb-info" tabIndex={-1} className="mt-8 bg-stone-50 print:hidden">
      <div>
        {!bilingual && (
          <div className="bg-gray-700 text-white">
            <section className="container py-6">
              <h2 className="mb-4">{t('gcweb:footer.service-canada')}</h2>
              <div className="grid gap-x-4 gap-y-2 text-sm sm:grid-cols-3">
                <AppLink
                  to={t('gcweb:app.contact-us.href', { baseUri: MSCA_BASE_URL })}
                  property="item"
                  typeof="WebPage"
                  className="hover:underline"
                >
                  <span property="name">{t('gcweb:app.contact-us')}</span>
                </AppLink>
              </div>
            </section>
          </div>
        )}
      </div>
      <div className="container flex items-center justify-between gap-6 py-2.5 sm:py-3.5">
        {bilingual ? (
          <h2 className="sr-only">
            <span lang="en">{en('gcweb:footer.about-site')}</span> / <span lang="fr">{fr('gcweb:footer.about-site')}</span>
          </h2>
        ) : (
          <h2 className="sr-only">{t('gcweb:footer.about-site')}</h2>
        )}
        <div className="flex flex-col items-start gap-2 text-sm leading-6 sm:flex-row sm:items-center sm:gap-4">
          <AppLink
            className="text-slate-700 hover:underline"
            to={t('gcweb:footer.terms-conditions.href', { baseUri: ECAS_BASE_URL })}
            data-gc-analytics-navigation={`Footer:Footer:${t('gcweb:footer.terms-conditions.text')}`}
          >
            {t('gcweb:footer.terms-conditions.text')}
          </AppLink>
          <div className="hidden size-0 rounded-full border-[3px] border-slate-700 sm:block"></div>
          <AppLink
            className="text-slate-700 hover:underline"
            to={t('gcweb:footer.privacy.href', { baseUri: ECAS_BASE_URL })}
            data-gc-analytics-navigation={`Footer:Footer:${t('gcweb:footer.privacy.text')}`}
          >
            {t('gcweb:footer.privacy.text')}
          </AppLink>
        </div>
        <div>
          <img
            src="https://www.canada.ca/etc/designs/canada/wet-boew/assets/wmms-blk.svg"
            alt={bilingual ? `${en('gcweb:footer.gc-symbol')} / ${fr('gcweb:footer.gc-symbol')}` : t('gcweb:footer.gc-symbol')}
            width={300}
            height={71}
            className="h-10 w-auto"
          />
        </div>
      </div>
    </footer>
  );
}
