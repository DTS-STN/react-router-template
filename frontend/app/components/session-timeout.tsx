import { useEffect, useState } from 'react';

import { useFetchers, useLocation, useNavigation } from 'react-router';

import { useTranslation } from 'react-i18next';
import type { IIdleTimerProps } from 'react-idle-timer';
import { useIdleTimer } from 'react-idle-timer';

import { Button } from '~/components/button';
import { Dialog, DialogContent, DialogFooter, DialogHeader, DialogTitle } from '~/components/dialog';

export interface SessionTimeoutProps extends Required<Pick<IIdleTimerProps, 'promptBeforeIdle' | 'timeout'>> {
  /**
   * Function to call when the session end is triggered.
   */
  onSessionEnd: () => Promise<void> | void;

  /**
   * Function to call when the session extend is triggered.
   */
  onSessionExtend: () => Promise<void> | void;
}

/**
 * SessionTimeout component to handle user session timeout prompts.
 * This component uses the `react-IdleTimer` library to manage idle time and prompt the user
 * before the session expires. It integrates with Remix's to activate the IdleTimer on route changes,
 * fetcher submissions, and form submissions.
 */
export function SessionTimeout({ promptBeforeIdle, timeout, onSessionEnd, onSessionExtend }: SessionTimeoutProps) {
  const { t } = useTranslation(['gcweb']);
  const [timeRemaining, setTimeRemaining] = useState('');

  const { activate, isPrompted, getRemainingTime } = useIdleTimer({
    // Disable default event listeners; The IdleTimer should only activate during route navigation and form
    // submissions, as these actions will interact with the session and extend its lifespan.
    events: [],
    onIdle: endSession,
    promptBeforeIdle,
    timeout,
  });

  const { key: locationKey } = useLocation();

  useEffect(() => {
    activate(); // Activate the IdleTimer whenever the location changes (indicating a route navigation).
  }, [locationKey, activate]);

  const fetchers = useFetchers();
  const fetcherSubmitting = fetchers.some(({ state }) => state === 'submitting');

  useEffect(() => {
    if (fetcherSubmitting) {
      activate(); // Activate the IdleTimer if any fetcher is submitting.
    }
  }, [fetcherSubmitting, activate]);

  const { state: navigationState } = useNavigation();
  const formSubmitting = navigationState === 'submitting';

  useEffect(() => {
    if (formSubmitting) {
      activate(); // Activate the IdleTimer if a <Form> is submitting.
    }
  }, [formSubmitting, activate]);

  useEffect(() => {
    const updateRemainingTime = () => {
      const remainingTime = getRemainingTime();
      const minutes = Math.floor(remainingTime / 60_000);
      const seconds = Math.floor((remainingTime % 60_000) / 1_000);
      const formattedTime = `${minutes}:${seconds.toString().padStart(2, '0')}`;
      setTimeRemaining(formattedTime);
    };

    const interval = setInterval(updateRemainingTime, 1000);
    updateRemainingTime(); // Initial call to set the time immediately

    return () => clearInterval(interval);
  }, [getRemainingTime]);

  async function endSession(): Promise<void> {
    await onSessionEnd();
    activate();
  }

  async function extendSession(): Promise<void> {
    await onSessionExtend();
    activate();
  }

  return (
    <Dialog open={isPrompted()} onOpenChange={(open) => !open && extendSession()}>
      <DialogContent aria-describedby={undefined} className="sm:max-w-md">
        <DialogHeader>
          <DialogTitle>{t('gcweb:session-timeout.header')}</DialogTitle>
        </DialogHeader>
        {t('gcweb:session-timeout.description', { timeRemaining })}
        <DialogFooter>
          <Button id="end-session-button" variant="default" size="sm" onClick={endSession}>
            {t('gcweb:session-timeout.end-session')}
          </Button>
          <Button id="continue-session-button" variant="primary" size="sm" onClick={extendSession}>
            {t('gcweb:session-timeout.continue-session')}
          </Button>
        </DialogFooter>
      </DialogContent>
    </Dialog>
  );
}
