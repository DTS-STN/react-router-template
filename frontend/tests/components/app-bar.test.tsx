import { createRoutesStub } from 'react-router';

import { render } from '@testing-library/react';
import { describe, expect, it } from 'vitest';
import { mock } from 'vitest-mock-extended';

import { AppBar } from '~/components/app-bar';

describe('AppBar', () => {
  it('should correctly render an AppBar with a MenuItem when the file property is provided', () => {
    globalThis.__appEnvironment = mock({
      MSCA_BASE_URL: 'https://msca.example.com/',
    });

    const RoutesStub = createRoutesStub([
      {
        path: '/fr/',
        Component: () => <AppBar />,
      },
    ]);

    const { container } = render(<RoutesStub initialEntries={['/fr/']} />);

    expect(container).toMatchSnapshot('expected html');
  });

  it('should correctly render an AppBar with a MenuItem when the to property is provided', () => {
    globalThis.__appEnvironment = mock({
      MSCA_BASE_URL: 'https://msca.example.com/',
    });

    const RoutesStub = createRoutesStub([
      {
        path: '/fr/',
        Component: () => <AppBar />,
      },
    ]);

    const { container } = render(<RoutesStub initialEntries={['/fr/']}></RoutesStub>);

    expect(container).toMatchSnapshot('expected html');
  });

  it('should render render an AppBar with a name provided', () => {
    globalThis.__appEnvironment = mock({
      MSCA_BASE_URL: 'https://msca.example.com/',
    });

    const RoutesStub = createRoutesStub([
      {
        path: '/fr/',
        Component: () => <AppBar />,
      },
    ]);

    const { container } = render(<RoutesStub initialEntries={['/fr/']}></RoutesStub>);

    expect(container).toMatchSnapshot('expected html');
  });

  it('should render render an AppBar with a profile item provided', () => {
    globalThis.__appEnvironment = mock({
      MSCA_BASE_URL: 'https://msca.example.com/',
    });

    const RoutesStub = createRoutesStub([
      {
        path: '/fr/',
        Component: () => <AppBar name="Test User" />,
      },
    ]);
    const { container } = render(<RoutesStub initialEntries={['/fr/']}></RoutesStub>);
    expect(container).toMatchSnapshot('expected html');
  });
});
