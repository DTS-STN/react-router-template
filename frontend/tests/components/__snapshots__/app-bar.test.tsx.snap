// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AppBar > should correctly render an AppBar with a MenuItem when the file property is provided > expected html 1`] = `
<div>
  <div
    class="bg-slate-700"
  >
    <div
      class="align-center container mx-auto flex flex-wrap justify-between"
    >
      <div
        class="align-center flex"
      >
        <span
          class="my-auto py-2 text-white sm:text-2xl"
          id="menu-label"
        >
          <a
            class="hover:underline"
            data-discover="true"
            href="/fr/{"key":"gcweb:app.menu-dashboard.href","options":{"baseUri":"https://msca.example.com/"}}"
          >
            gcweb:app.title
          </a>
        </span>
      </div>
      <div
        class="flex items-center space-x-4 text-right"
      />
    </div>
  </div>
</div>
`;

exports[`AppBar > should correctly render an AppBar with a MenuItem when the to property is provided > expected html 1`] = `
<div>
  <div
    class="bg-slate-700"
  >
    <div
      class="align-center container mx-auto flex flex-wrap justify-between"
    >
      <div
        class="align-center flex"
      >
        <span
          class="my-auto py-2 text-white sm:text-2xl"
          id="menu-label"
        >
          <a
            class="hover:underline"
            data-discover="true"
            href="/fr/{"key":"gcweb:app.menu-dashboard.href","options":{"baseUri":"https://msca.example.com/"}}"
          >
            gcweb:app.title
          </a>
        </span>
      </div>
      <div
        class="flex items-center space-x-4 text-right"
      />
    </div>
  </div>
</div>
`;

exports[`AppBar > should render render an AppBar with a name provided > expected html 1`] = `
<div>
  <div
    class="bg-slate-700"
  >
    <div
      class="align-center container mx-auto flex flex-wrap justify-between"
    >
      <div
        class="align-center flex"
      >
        <span
          class="my-auto py-2 text-white sm:text-2xl"
          id="menu-label"
        >
          <a
            class="hover:underline"
            data-discover="true"
            href="/fr/{"key":"gcweb:app.menu-dashboard.href","options":{"baseUri":"https://msca.example.com/"}}"
          >
            gcweb:app.title
          </a>
        </span>
      </div>
      <div
        class="flex items-center space-x-4 text-right"
      />
    </div>
  </div>
</div>
`;

exports[`AppBar > should render render an AppBar with a profile item provided > expected html 1`] = `
<div>
  <div
    class="bg-slate-700"
  >
    <div
      class="align-center container mx-auto flex flex-wrap justify-between"
    >
      <div
        class="align-center flex"
      >
        <span
          class="my-auto py-2 text-white sm:text-2xl"
          id="menu-label"
        >
          <a
            class="hover:underline"
            data-discover="true"
            href="/fr/{"key":"gcweb:app.menu-dashboard.href","options":{"baseUri":"https://msca.example.com/"}}"
          >
            gcweb:app.title
          </a>
        </span>
      </div>
      <div
        class="flex items-center space-x-4 text-right"
      >
        <button
          aria-expanded="false"
          aria-haspopup="menu"
          class="flex h-full flex-nowrap space-x-2 bg-slate-200 px-2 text-sm text-slate-700 hover:bg-neutral-300 hover:underline focus:bg-neutral-300 focus:ring-2 focus:ring-black focus:ring-offset-2 focus:outline-hidden aria-expanded:bg-neutral-300 aria-expanded:text-slate-700 sm:space-x-4 sm:px-4"
          data-state="closed"
          id="radix-«r0»"
          type="button"
        >
          <div
            class="text-md my-auto flex flex-nowrap items-center space-x-2 py-2"
          >
            <div
              class="flex h-8 w-8 items-center justify-center rounded-full bg-slate-700"
            >
              <svg
                aria-hidden="true"
                class="svg-inline--fa fa-user size-5 text-slate-200"
                data-icon="user"
                data-prefix="fas"
                focusable="false"
                role="img"
                viewBox="0 0 448 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M224 256A128 128 0 1 0 224 0a128 128 0 1 0 0 256zm-45.7 48C79.8 304 0 383.8 0 482.3C0 498.7 13.3 512 29.7 512l388.6 0c16.4 0 29.7-13.3 29.7-29.7C448 383.8 368.2 304 269.7 304l-91.4 0z"
                  fill="currentColor"
                />
              </svg>
            </div>
            <span
              class="text-md hidden py-2 font-bold sm:block"
              id="menu-label"
            >
              Test User
            </span>
          </div>
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-chevron-down my-auto size-4"
            data-icon="chevron-down"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 512 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M233.4 406.6c12.5 12.5 32.8 12.5 45.3 0l192-192c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0L256 338.7 86.6 169.4c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3l192 192z"
              fill="currentColor"
            />
          </svg>
        </button>
      </div>
    </div>
  </div>
</div>
`;
